src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -437,51 +464,143 @@ public void delete(String modelId, ActionListener<DeleteModelResponse> listener)
 
-            // On model deletion from the index, remove the model from all nodes' model cache
-            ActionListener<DeleteResponse> onModelDeleteListener = ActionListener.wrap(deleteResponse -> {
-                // If model is not deleted, return with error message
+            // On model metadata removal, delete the model from the index
+            clearModelMetadataStep.whenComplete(
+                acknowledgedResponse -> deleteModelFromIndex(modelId, deleteModelFromIndexStep, deleteRequestBuilder),
+                listener::onFailure
+            );
+
+            deleteModelFromIndexStep.whenComplete(deleteResponse -> {
+                // If model is not deleted, remove modelId from model graveyard and return with error message
                 if (deleteResponse.getResult() != DocWriteResponse.Result.DELETED) {
+                    updateModelGraveyardToDelete(modelId, true, unblockModelIdStep);
                     String errorMessage = String.format("Model \" %s \" does not exist", modelId);
                     listener.onResponse(new DeleteModelResponse(modelId, deleteResponse.getResult().getLowercase(), errorMessage));
                     return;
                 }
 
-                // After model is deleted from the index, make sure the model is evicted from every cache in the
-                // cluster
-                client.execute(
-                    RemoveModelFromCacheAction.INSTANCE,
-                    new RemoveModelFromCacheRequest(modelId),
-                    ActionListener.wrap(removeModelFromCacheResponse -> {
+                // After model is deleted from the index, make sure the model is evicted from every cache in the cluster
+                removeModelFromCache(modelId, clearModelFromCacheStep);
+            }, e -> listener.onResponse(new DeleteModelResponse(modelId, "failed", e.getMessage())));
 
-                        if (!removeModelFromCacheResponse.hasFailures()) {
-                            listener.onResponse(new DeleteModelResponse(modelId, deleteResponse.getResult().getLowercase(), null));
-                            return;
-                        }
+            clearModelFromCacheStep.whenComplete(removeModelFromCacheResponse -> {
+                // Remove modelId from model graveyard
+                updateModelGraveyardToDelete(modelId, true, unblockModelIdStep);
 
-                        String failureMessage = buildRemoveModelErrorMessage(modelId, removeModelFromCacheResponse);
+                unblockModelIdStep.whenComplete(acknowledgedResponse -> {
 
-                        listener.onResponse(new DeleteModelResponse(modelId, "failed", failureMessage));
+                    // After clearing the cache, if there are no errors return the response
+                    if (!removeModelFromCacheResponse.hasFailures()) {
+                        listener.onResponse(new DeleteModelResponse(modelId, "deleted", null));
```

Comment:
same as for "failed", please make it a constant

Reply:
ack

Reply:
@[naveentatikonda](https://github.com/naveentatikonda) did you make a constant for "deleted" result code? I do see only one for failed.

Reply:
Sorry, my bad. Updated it now.